storage/innobase/handler/ha_innodb.cc

```diff
@@ -17080,6 +17080,52 @@ int innobase_rollback_by_xid(handlerton* hton, XID* xid)
 	}
 }
 
+/**
+This function is used to rollback one X/Open XA distributed transaction
+which is in the prepared state asynchronously.
+
+It only set the transaction's status to ACTIVE and persist the status.
+The transaction will be rolled back by background rollback thread.
+
+@param[in] hton InnoDB handlerton
+@param[in] xid X/Open XA transaction identification
+
+@return 0 or error number */
+int innobase_recover_rollback_by_xid(handlerton *hton, XID *xid)
+{
+	DBUG_ASSERT(hton == innodb_hton_ptr);
+	DBUG_EXECUTE_IF("innobase_xa_fail", return XAER_RMFAIL;);
+	DBUG_EXECUTE_IF("crash_before_async_binlog_recovery", DBUG_SUICIDE(););
```

Comment:
I do not see any test coverage of this crash injection point.

I would strongly prefer to use `DEBUG_SYNC` for any crash injection. Simply make the server hang on
```sql
SET DEBUG_SYNC='some_label WAIT_FOR ever';
```
and let the test client kill and restart the server. In that way, it is readily compatible with Valgrind.

Reply:
right, removed 'crash_before_async_binlog_recovery'.